### PR TITLE
Extend live ranges for values defined outside of the loop

### PR DESCRIPTION
**Pull request by Krovatkin**

### CODE REVIEW

**Comment by Krovatkin** [comment]
fixes https://github.com/pytorch/pytorch/issues/30494

File: torch/csrc/jit/passes/liveness.cpp [hunk 1]
```diff
@@ -88,6 +97,9 @@ struct LivenessAnalyzer {
         loop_block = processBlock(it->blocks()[0], loop_block);
         // loop block's inputs die outside loop's block
         loop_block -= toSparseBitVector(it->blocks()[0]->inputs());
+        // everything that is alive at the beginning of the loop
+        // needs to be extended until the end of the loop
+        extendLiveness(it->blocks()[0], loop_block);
```
Comment (zdevito) [inline torch/csrc/jit/passes/liveness.cpp]: Liveness analysis is a classic dataflow analysis, and they all require iterating until convergence on a loop. I don't see such a convergence iteration here so I believe this is probably still incorrect and only pushes the bug down to a more complicated example.

Comment (Krovatkin) [reply torch/csrc/jit/passes/liveness.cpp]: @zdevito added a comment

**Review by zdevito — COMMENTED** [review]
Is it possible to make the explanation self contained?  Without reading the paper, it is hard to understand.  The paper itself has a lot of stuff it in that is slightly different than our system, making it hard to figure out how to translate its explanation into terms familar to the PyTorch IR.

**Comment by Krovatkin** [comment]
@zdevito I restructured liveness so it follows the algorithms described in the paper very closely. I tried to summarize their algorithms using our terms and include the terms used in the paper in case someone wants to read the paper.

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 2154c87:


* **2/2** broken upstream at merge base 540b9da (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=540b9da41e87473948bb2868c2da899900511c39&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=2154c87e92265c0bc7463c1b54e3f25c6999ce38).

### :construction: 2 upstream failures recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_macos_10_13_py3_test](https://dr.pytorch.org/build-details.html?build_id=77785995)
* [pytorch_linux_xenial_py3_clang5_asan_test](https://dr.pytorch.org/build-details.html?build_id=77785996)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Review by zdevito — COMMENTED** [review]
It would help a lot if the explanation was self contained. That is, it should not require a reader of the code to also read a multi-page paper about generic liveness to figure out why this is correct.  The code and the comment both say what is being doing, but are largely missing the intuition for why this is sufficient.

**Comment by Krovatkin** [comment]
https://github.com/pytorch/pytorch/pull/30643/files#diff-35b45c6cc8bc8efb496a3df5f3e573deR36 @zdevito I also tried to summarize the intuition.

**Comment by Krovatkin** [comment]
fixed in https://github.com/pytorch/pytorch/pull/31724

**Pull request closed without merging**